test/server/shared/root-hooks.mjs

```diff
@@ -0,0 +1,112 @@
+import { spawn } from 'child_process';
+import net from 'node:net';
+import sinon from 'sinon';
+import { promisify } from 'node:util';
+import { server as config } from '../../../app/models/config-model.js';
+
+/**
+ * @return {Promise<number>}
+ */
+const getAvailablePort = () => {
+    const server = net.createServer();
+
+    return new Promise((resolve, reject) => {
+        server.listen(0, () => {
+            const { port } = server.address();
+
+            server.close((error) => {
+                if (error) {
+                    reject(error);
+                } else {
+                    resolve(port);
+                }
+            });
+        });
+
+        server.once('error', (error) => {
+            server.close(() => {
+                reject(error);
+                server.unref();
+            });
+        });
+    });
+};
+
+/**
+ * @type {number} port
+ */
+const startRedisServer = async (port) =>
+    new Promise((resolve, reject) => {
+        let isSettled = false;
+
+        const server = spawn('redis-server', ['--port', String(port)]);
+
+        server.stdout.once('data', () => {
+            if (!isSettled) {
+                resolve(server);
+                isSettled = true;
+            }
+        });
+
+        server.once('spawn', () => {
+            if (!isSettled) {
+                resolve(server);
+                isSettled = true;
+            }
+        });
+
+        server.once('error', (error) => {
+            if (!isSettled) {
+                reject(error);
+                isSettled = true;
+            }
+        });
+
+        return server;
+    });
+
+const [mainPort, cachePort] = await Promise.all([
+    getAvailablePort(),
+    getAvailablePort(),
+]);
+
+const redisServerProcesses = await Promise.all([
+    startRedisServer(mainPort),
+    startRedisServer(cachePort),
+]);
+
+const sandbox = sinon.createSandbox();
+
+sandbox.stub(config.redis.main, 'port').value(mainPort);
+sandbox.stub(config.redis.cache, 'port').value(cachePort);
```

Comment:
I'm hoping mocha runs this before requiring all the other source files :+1:

Reply:
I validated that it does, after accidentally wiping my dev databases several times. I'll add a comment clarifying how/why it works so it doesn't get broken by mistake.